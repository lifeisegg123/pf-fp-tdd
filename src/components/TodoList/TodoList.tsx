import { Todo } from "src/interfaces";
import { TodoItem } from "../TodoItem";

interface TodoListProps {
  todos: Todo[];
  onCheck: (id: string) => () => void;
  onRemove: (id: string) => () => void;
}

export function TodoList({ todos, onCheck, onRemove }: TodoListProps) {
  return (
    <ul>
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          {...todo}
          onCheck={onCheck(todo.id)}
          onRemove={onRemove(todo.id)}
        />
      ))}
    </ul>
  );
}
