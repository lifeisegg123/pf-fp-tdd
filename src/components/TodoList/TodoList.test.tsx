import { TodoList } from "./TodoList";
import { fireEvent, render, screen } from "@testing-library/react";
import { useState } from "react";
import { filter, map, pipe, toArray } from "@fxts/core";

const MOCK_TODOS = [
  { id: "1", desc: "test1", checked: false },
  { id: "2", desc: "test2", checked: false },
  { id: "3", desc: "test3", checked: false },
  { id: "4", desc: "test4", checked: false },
  { id: "5", desc: "test5", checked: false },
  { id: "6", desc: "test6", checked: false },
];

const MockTodo = () => {
  const [todos, setTodos] = useState(MOCK_TODOS);

  const onRemove = (targetId: string) => () => {
    setTodos((prev) =>
      pipe(
        prev,
        filter(({ id }) => targetId !== id),
        toArray
      )
    );
  };
  const onCheck = (targetId: string) => () => {
    setTodos((prev) =>
      pipe(
        prev,
        map(({ id, checked, desc }) => ({
          id,
          checked: id === targetId ? !checked : checked,
          desc,
        })),
        toArray
      )
    );
  };
  return <TodoList onCheck={onCheck} onRemove={onRemove} todos={todos} />;
};

describe("Component/TodoList", () => {
  it("should match snapshot", () => {
    const { container } = render(<MockTodo />);
    expect(container).toMatchSnapshot();
  });

  it("should remove item", () => {
    const { getByText } = render(<MockTodo />);

    const target = getByText("test1").nextSibling;
    fireEvent.click(target as HTMLButtonElement);

    expect(target).not.toBeInTheDocument();
  });

  it("should handle check", () => {
    const { getByText } = render(<MockTodo />);

    const target = getByText("test1");

    fireEvent.click(target);

    expect(screen.getByText("done")).toBeInTheDocument();
  });
});
