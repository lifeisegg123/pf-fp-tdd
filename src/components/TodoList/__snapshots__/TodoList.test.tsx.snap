// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/TodoList should match snapshot 1`] = `
<div>
  <ul>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test1
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test2
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test3
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test4
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test5
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
    <li
      class="flex-row-center gap-sm"
    >
      <span>
        todo
      </span>
      <h2>
        test6
      </h2>
      <button
        aria-label="삭제"
      >
        X
      </button>
    </li>
  </ul>
</div>
`;
